helix-core/src/main/java/org/apache/helix/api/rebalancer/waged/constraints/HardConstraint.java

```diff
@@ -0,0 +1,50 @@
+package org.apache.helix.api.rebalancer.waged.constraints;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.helix.controller.rebalancer.waged.model.AssignableNode;
+import org.apache.helix.controller.rebalancer.waged.model.AssignableReplica;
+import org.apache.helix.controller.rebalancer.waged.model.ClusterContext;
+
+/**
+ * Evaluate a partition allocation proposal and return YES or NO based on the cluster context.
+ * Any proposal fails one or more hard constraints will be rejected.
+ */
+public interface HardConstraint {
+  enum FailureReason {
+    FAULT_ZONES_CONTAIN_SAME_PARTITION,
+    NODES_DEACTIVATED,
+    NODES_NO_TAG,
+    NODES_EXCEED_MAX_PARTITION,
+    NODES_INSUFFICIENT_RESOURCE,
+    NODES_CONTAIN_SAME_PARTITION,
+  }
+
+  /**
+   * @return True if the proposed assignment is valid.
+   */
+  boolean isAssignmentValid(AssignableNode node, AssignableReplica rep,
+      ClusterContext clusterContext);
+
+  /**
+   * @return Detail of the reason that the proposed assignment was rejected.
```

Comment:
Same comments like above. We should create a java class that encapsulates the status, as well as the string explanation at the same time, clients only need to call one single method.

`class Status {
    boolean isValid;
    FailureReason reason; // if isValid == false, else null or empty string.
}`

Reply:
This is different from the previous one. The failure reason for each HardConstraint implementation is fixed. Leave this method here is just for convenience. So the rebalancer can gather information easier.